include/fraction.h

```diff
@@ -0,0 +1,195 @@
+
+#include <cassert>
+#include <numeric>
+
+// Class to represent simple fractions. The fraction is always simplified
+// after construction or after any operation. The sign always is normalised so
+// the numerator carries the sign and the denominator is always positive.
+class Fraction {
+public:
+	constexpr Fraction()
+	{
+		num   = 0;
+		denom = 1;
+	}
+
+	constexpr Fraction(const int64_t n) : num(n), denom(1) {}
+
+	constexpr Fraction(const int64_t n, const int64_t d) : num(n), denom(d)
+	{
+		assert(d != 0);
+
+		if (d == 0 || n == 0) {
+			num   = 0;
+			denom = 1;
+		} else {
+			num   = n;
+			denom = d;
+
+			// Simplify fraction
+			auto gcd = std::gcd(num, denom);
+
+			num /= gcd;
+			denom /= gcd;
+
+			// Normalise sign so the denominator is always positive
+			int sign = 1;
+			if (num < 0) {
+				sign *= -1;
+				num *= -1;
+			}
+			if (denom < 0) {
+				sign *= -1;
+				denom *= -1;
+			}
+
+			num *= sign;
+		}
+	}
+
+	constexpr int64_t Num() const
+	{
+		return num;
+	}
+
+	constexpr int64_t Denom() const
+	{
+		return denom;
+	}
+
+	constexpr Fraction Inverse() const
+	{
+		return Fraction(denom, num);
+	}
+
+	constexpr double ToDouble() const
+	{
+		return static_cast<double>(num) / denom;
+	}
+
+	constexpr double ToFloat() const
+	{
+		return static_cast<float>(num) / denom;
+	}
+
+	// Equality
+	constexpr bool operator==(const Fraction& that) const
+	{
+		return (num == that.num) && (denom == that.denom);
+	}
+
+	constexpr bool operator!=(const Fraction& that) const
+	{
+		return !operator==(that);
+	}
+
+	// Addition
+	constexpr Fraction operator+=(const Fraction& that)
+	{
+		*this = {num * that.denom + that.num * denom, denom * that.denom};
+		return *this;
+	}
+
+	constexpr Fraction operator+(const Fraction& that) const
+	{
+		auto tmp = *this;
+		tmp += that;
+		return tmp;
+	}
+
+	// Subtraction
+	constexpr Fraction operator-=(const Fraction& that)
+	{
+		*this = {num * that.denom - that.num * denom, denom * that.denom};
+		return *this;
+	}
+
+	constexpr Fraction operator-(const Fraction& that) const
+	{
+		auto tmp = *this;
+		tmp -= that;
+		return tmp;
+	}
+
+	// Multiplication
+	constexpr Fraction& operator*=(const int64_t s)
+	{
+		*this = {num * s, denom};
+		return *this;
+	}
+
+	constexpr Fraction operator*(const int64_t s) const
+	{
+		auto tmp = *this;
+		tmp *= s;
+		return tmp;
+	}
+
+	constexpr Fraction& operator*=(const Fraction& that)
+	{
+		*this = {num * that.num, denom * that.denom};
+		return *this;
+	}
+
+	constexpr Fraction operator*(const Fraction& that) const
+	{
+		auto tmp = *this;
+		tmp *= that;
+		return tmp;
+	}
+
+	// Division
+	constexpr Fraction operator/=(const int64_t s)
+	{
+		*this = {num, denom * s};
+		return *this;
+	}
+
+	constexpr Fraction operator/(const int64_t s) const
+	{
+		auto tmp = *this;
+		tmp /= s;
+		return tmp;
+	}
+
+	constexpr Fraction operator/=(const Fraction& that)
+	{
+		*this = {num * that.denom, denom * that.num};
+		return *this;
+	}
```

Comment:
Yeah, will rewrite as suggested. But... I'd fully expect the compiler to figure all that out. My general view is to only care this low level drudgery where it matters. I thinks that's less than 10% of the total code.  Otherwise copies all the way, no?

Reply:
For the record, the (much less than) 10% of the code that actually _is_ performance intensive in this PR is the bilinear upscaler, I made that branchless; probably it doesn't get any better than that. Plus the pixel-based operations (decoding and sRGB<->linear conversion). Those benefit from micro-optimisations and I've done my best. The rest are macro-level optimisations (e.g. only do line-based encoding, don't reallocate buffers or make unnecessary copies of image data, etc.)

I guess what I'm saying we can default to sane optimal code if it's easy enough as a default, but in truth, there's a few places where it really matter. So I usually go with what's easiest 😄

Reply:
Refactored.